keps/sig-node/4753-expose-ownerref-downard-api/README.md

```diff
@@ -0,0 +1,616 @@
+but to extend the existing `downward API` to include the `metadata.ownerReferences` field.
+
+### Implementation Details
+
+* Will use pre-existing information from the Kubelet and VolumeHost structs
+to extract pod metadata information. 
+
+At `/pkg/apis/core/validation/validation.go` we will edit the following:
+```go
+var validVolumeDownwardAPIFieldPathExpressions = sets.New(
+	"metadata.name",
+	"metadata.namespace",
+	"metadata.labels",
+	"metadata.annotations",
+	"metadata.uid",
+  "metadata.ownerReferences")
+```
+
+At `pkg/apis/core/pods/helpers.go` we will edit the following function:
+
+```go
+func ConvertDownwardAPIFieldLabel(version, label, value string) (string, string, error) {
+	if version != "v1" {
+		return "", "", fmt.Errorf("unsupported pod version: %s", version)
+	}
+
+	if path, _, ok := fieldpath.SplitMaybeSubscriptedPath(label); ok {
+		switch path {
+		case "metadata.annotations", "metadata.labels":
+			return label, value, nil
+		default:
+			return "", "", fmt.Errorf("field label does not support subscript: %s", label)
+		}
+	}
+
+	switch label {
+	case "metadata.annotations",
+		"metadata.labels",
+		"metadata.name",
+		"metadata.namespace",
+		"metadata.uid",
+    "metadata.ownerReferences",
+		"spec.nodeName",
+		"spec.restartPolicy",
+		"spec.serviceAccountName",
+		"spec.schedulerName",
+		"status.phase",
+		"status.hostIP",
+		"status.hostIPs",
+		"status.podIP",
+		"status.podIPs":
+		return label, value, nil
+	// This is for backwards compatibility with old v1 clients which send spec.host
+	case "spec.host":
+		return "spec.nodeName", value, nil
+	default:
+		return "", "", fmt.Errorf("field label not supported: %s", label)
+	}
+}
+```
+* Information will be passed to the pod via the `valueFrom` field in the volume.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: kubernetes-downwardapi-volume-example
+  ownerReferences:
+  - apiVersion: apps/v1
+    blockOwnerDeletion: true
+    controller: true
+    kind: DaemonSet
+    name: an-owned-pod-1722852739
+    uid: 6be1683f-da9c-4f68-9440-82376231cfa6
+  - apiVersion: apps/v1
+    blockOwnerDeletion: true
+    controller: true
+    kind: Some-CRD
+    name: an-owned-pod-1722852739
+    uid: 6be1683f-da9c-4f68-9440-82376231cfa6
+spec:
+  containers:
+    - name: client-container
+      image: busybox
+      command: ["sh", "-c"]
+      args:
+      - while true; do
+          if [[ -e /etc/podinfo/ownerReferences ]]; then
+            echo -en '\n\n'; cat /etc/podinfo/ownerReferences; fi;
+          sleep 5;
+        done;
+      volumeMounts:
+        - name: podinfo
+          mountPath: /etc/podinfo
+  volumes:
+    - name: podinfo
+      downwardAPI:
+        items:
+          - path: "ownerReferences"
+            fieldRef:
+              fieldPath: metadata.ownerReferences
+```
+
+Logs from the pod will show the ownerReferences of the pod:
+
+```bash
+$ kubectl logs kubernetes-downwardapi-volume-example
```

Comment:
This is under-specified.

I don't know if we should dump the WHOLE object, but it's probably fine - can you think of any reason to hide any of these?

YAML is a terrible file format for parsing - I think it would be better to spec this as something like:

"""
The file representation of this fieldref is a JSON list of objects, each of which is a serialized `k8s.io/apimachinery/pkg/apis/meta/v1.OwnerReference`.
"""

Open question:  Should we instead produce a metav1.List?  e.g.

```
{
    "kind": "OwnerReference",
    "apiVersion": "meta/v1",
    "items": [
        { .... }, { ... }
    ]
}
```

IF there was ever a meta/v2, we would need to allow people to request that - does the output need to be self-documenting?  I think probably not.

Reply:
> I don't know if we should dump the WHOLE object, but it's probably fine - can you think of any reason to hide any of these?

I think if we go for your JSON recommendation and only pass a list is more than ok, all we need is the ownerRef to pass it to objects that the Pod will create. 

So to your open question, I like the `metav1.List` idea.